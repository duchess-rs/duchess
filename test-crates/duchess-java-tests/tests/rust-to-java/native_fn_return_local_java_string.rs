use duchess::{java, prelude::*};

duchess::java_package! {
    package native_greeting;

    public class native_greeting.Native {
        public native_greeting.Native();
        public java.lang.String greet(java.lang.String);
        native java.lang.String baseGreeting(java.lang.String);
    }
}

#[duchess::java_function(native_greeting.Native::baseGreeting)]
fn base_greeting<'n>(
    _this: &native_greeting::Native,
    name: &'n java::lang::String,
) -> duchess::Result<duchess::Local<'n, java::lang::String>> {
    name.execute() //~ ERROR: trait bound
}

fn main() -> duchess::Result<()> {
    duchess::Jvm::builder()
        .link(base_greeting::java_fn())
        .try_launch()?;

    let n: String = native_greeting::Native::new()
        .greet("Ferris")
        .assert_not_null()
        .execute()
        .unwrap();

    assert_eq!(n, "Ferris, from Java");

    Ok(())
}
