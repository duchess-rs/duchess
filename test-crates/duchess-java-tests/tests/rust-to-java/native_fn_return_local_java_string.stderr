error[E0277]: the trait bound `&duchess::java::lang::String: IntoRust<Local<'_, duchess::java::lang::String>>` is not satisfied
   --> tests/ui/native_fn_return_local_java_string.rs:18:10
    |
18  |     name.execute()
    |          ^^^^^^^ the trait `IntoRust<Local<'_, duchess::java::lang::String>>` is not implemented for `&duchess::java::lang::String`
    |
    = help: the trait `IntoRust<std::string::String>` is implemented for `&duchess::java::lang::String`
    = help: for that trait implementation, expected `std::string::String`, found `Local<'_, duchess::java::lang::String>`
note: required by a bound in `execute`
   --> /home/ubuntu/crates/duchess/src/jvm.rs:103:39
    |
101 |     fn execute<R>(self) -> crate::Result<R>
    |        ------- required by a bound in this associated function
102 |     where
103 |         for<'jvm> Self::Output<'jvm>: IntoRust<R>,
    |                                       ^^^^^^^^^^^ required by this bound in `JvmOp::execute`

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0277`.
